src/hidpp20.c

```diff
@@ -2412,92 +2421,67 @@ hidpp20_onboard_profiles_parse(struct hidpp20_device *device,
 	crc_valid = hidpp20_onboard_profiles_is_sector_valid(device,
 							     profiles->sector_size,
 							     data);
-	if (crc_valid || !check_crc) {
+	if (crc_valid) {
 		for (i = 0; i < profiles->num_profiles; i++) {
 			uint8_t *d = data + 4 * i;
 
-			if (d[0] == 0xFF && d[1] == 0xFF)
+			if (d[HIDPP20_PROFILE_DIR_END1] == HIDPP20_PROFILE_DIR_END_VALUE &&
+			    d[HIDPP20_PROFILE_DIR_END2] == HIDPP20_PROFILE_DIR_END_VALUE)
 				break;
 
 			profiles->profiles[i].address = hidpp_get_unaligned_be_u16(d);
-			if (profiles->profiles[i].address != (dict_address | (i + 1)))
+
+			/* profile address sanity check */
+			if (profiles->profiles[i].address != (HIDPP20_USER_PROFILES_G402 | (i + 1)))
 				hidpp_log_info(&device->base,
-						"profile %d: error in the address: 0x%04x instead of 0x%04x\n",
-						i + 1,
-						profiles->profiles[i].address,
-						dict_address | (i + 1));
-			profiles->profiles[i].enabled = !!d[2];
+					       "profile %d: error in the address: 0x%04x instead of 0x%04x\n",
+					       i + 1,
+					       profiles->profiles[i].address,
+					       HIDPP20_USER_PROFILES_G402 | (i + 1));
+
+			profiles->profiles[i].enabled = !!d[HIDPP20_PROFILE_DIR_ENABLED];
 		}
+	} else {
+		hidpp_log_debug(device, "Profile directory has an invalid CRC... Reading ROM profiles.\n");
+
+		read_userdata = false;
```

Comment:
2/3: ... we are accessing the ROM profiles, which is good, ...

Reply:
Same. I don't understand what you want me to do :)